### ISSUE TITLE
Single/Multiple Windows accounts

### ISSUE THREAD

**Issue by AntonyCorbett** [AntonyCorbett/OnlyM#66]
From @DanJermy

**Should we have separate windows logins for each congregation?**

We have 2 congregations (plus a group that meets separately on occasion) that share the hall, and with our previous PC build using SoundBox we had separate windows logins separating configuration for soundbox (though I know that could have been done with soundbox presenting a list on launch), as well as for KHconf (which can also support multiple congs through a list), however we had dropbox synching with different accounts and the clear delineation of a windows account (with different PINs) made user error in selecting wrong config more unlikely (some soundbox config selected by each cong varied).

As part of the new W10 build with JWL in the brave new world, we've set up a folder structure in a single OneDrive account to store Recordings (with cong subfolder) and allow Uploading of media by any of our brothers in advance of the meetings - also in subfolders.

@AntonyCorbett  I notice your suggestion of the use of the 'id' parameter; and so far I've used the 'source' option under separate shortcuts for each congregation, as well as a third for a 'shared' folder where content might be needed by both congregations (i.e. for a local needs video for all congs rather than a picture for a public talk to be shown for just one cong)...  However, this will result in 3 desktop shortcuts for **OnlyM** alone!

I could do the same with **OnlyR** although I'd need to use additional folder junctions to redirect to the OneDrive path.  Alternatively it would be ideal if **OnlyR** could use a different folder path based on the time (as suggested in an existing feature request, such that it would use the timings set for the countdown timer in OnlyT).

**OnlyV** don't really require cong-specific config at this stage; **OnlyT** can accommodate multiple congs countdown timer already which is sufficient (although I could see opinions dividing on some of the other options); so I just need to determine the merits of separate logins for clarity (i.e. single icons to launch **OnlyM** and **OnlyR** with default paths being set, along with potential future scalability) _vs_ a single login which can be more consistent across both congregations.

If functionality was available to present a cong menu for each app or allow to be auto-selected by the time of day (and later overridden if necessary) then having a single login for all congs would seem more straight forward.  Would having an independent "OnlyConfig" app that the other apps reference to set these times and default folder locations defeat the purpose of separating the apps out?

Apologies for rambling nature of this post; hopefully enough to give folks a background on requirements such that I can pilfer the no-doubt well thought out implementations of other fellow KH techies out there.  Likely I'm over-complicating it, so help me to introduce the KISS principle where it applies.

Thanks in advance / Dan

_ps. bonus question - {Not sure if there's a more appropriate forum to ask overarching questions beyond the scope of an individual app or ones that are to do with JWL operation, but may I ask} Do most people use a standard account rather than an admin one for day to day Sound Desk operations?_

_Originally posted by @DanJermy in https://github.com/AntonyCorbett/OnlyM/issues/53#issuecomment-433748704_

**Comment by AntonyCorbett** [comment]
@DanJermy you could use the Windows 10 multiple desktop feature to organise the shortcuts, but my personal preference would be to have separate logins. This allows each congregation to take full responsibility for its content and settings.

**Comment by blazer2004** [comment]
> This allows each congregation to take full responsibility for its content and settings.
We take the oppisite approach. Our congregations use a common desktop (and a commen desktop computer, for that matter) and a common media folder (videos/jw library) for the meetings. It works out really well. Of course, a thought out wiring infrastructure is essental for the video equipment and the computer to post in the correct order every time. The desktop gets used as a temporary file storage area for pictures and extra videos, which each congregation respects. Longer term storage for videos that can't be downloaded via JW Library either gets placed in a special folder in videos, or the JW Library folder it's self. It then gets copied to the desktop the week before it's used.

**Comment by JohannesKreuzer** [comment]
We are doing it a similar way as @blazer2004.
We have one common user account for all congregations in our hall. Configuration of every important program is getting reset to a default config after every reboot via a autorunscript. I'm also working on a extension of the script to prompt for the congregation at startup and set the Meetings View of JWLibrary to the correct language based on the congregation now using the computer and also starting OnlyR with the correct parameter. The script is also deleting all files on the Desktop and in the Downloads folder so PC is always in a well defined state.
Doing it this way it is much less work than having to maintain 7 user accounts on two computers (we have a hall with two auditoriums and every congregation is able to use both halls, e.g. for mid week meeting at circuit overseer visit).

**Comment by DanJermy** [comment]
Thanks everyone for your help.

I think it may come down to a technical limitation of JW Library  -it doesn't seem to recognise the contents of the Folder junctions I created to use a single data source for downloaded music and videos. 

The junctions create fine but regardless of whether the next congregation 'user' account I create is an admin or standard user account or if it has full control permissions on the folder, when I launch JW Library it doesn't detect that those files are available to it.

Am I missing something or do all you folks that use separate user accounts for different congregations duplicate downloads by having each congregation download what they need?

**Speedy responses would be really appreciated as I'm attempting to implement this now (and therefore having to rethink my deployment approach of the OnlyM 'congregation folders' if we're to use a single login; i.e. I didn't really want lots of shortcuts on a shared desktop).**

**Comment by DanJermy** [comment]
So it seems JWL must hold a manifest somewhere determining which media files have been downloaded; It then requires connection to the internet to allow a 'download' however it doesn't seem to overwrite a song or video if it already exists (at least the modified date on the file doensn't change);

Once that user account knows it's 'downloaded' the file then it will access the file from the shared location via the junction, but it seems the manifest is what's causing the issue;  As a result I'm concerned that this likely unsupported configuration will present challenges in the future;  Anyone else have more info on this?

**Comment by AntonyCorbett** [comment]
@DanJermy JWL media content is managed by its database - a SqlLite database file that maintains an inventory of all available publications, media etc, and a per-user record of downloaded items, etc

**Comment by AntonyCorbett** [comment]
The JWL terms and conditions preclude reverse engineering or modification of this database,

**Comment by AntonyCorbett** [comment]
It should be possible to specify a single common source for downloads (for all user accounts). Then when an individual account is used, the JWL database will not immediately recognise that an item is already available. But when you click to download, that should be sufficient for JWL to find the existing media item and generate the required database entry.

**Comment by DanJermy** [comment]
@AntonyCorbett Understood - too ambitious then, thanks for confirming; shame there isn't a way to use a shared data repository for downloads; 
So options are...
(1) Have separate logins with folder junctions (which may carry risk of an unknown issue in the future) but the congregations need to still download their own files for each meeting;
(2) Have a single login for all users & need to have multiple desktop icons for OnlyM and OnlyR that specify the path for each congregation (carries risk that users will open the wrong ones);

Two ways to mitigate option 2 downsides that I can think of:
- add a congregation-list selection box on launch like Soundbox did and marry that to a different working folder path
- allow a meeting time-window setting so that the correct congregation (and therefore working folder path) is correctly determined by the time/day.

**Comment by DanJermy** [comment]
>           It should be possible to specify a single common source for downloads (for all user accounts). Then when an individual account is used, the JWL database will not immediately recognise that an item is already available. But when you click to download, that should be sufficient for JWL to find the existing media item and generate the required database entry.

I agree - but that's not possible at the moment right?

**Comment by DanJermy** [comment]
> The data store is in each user's AppData folder. There is no way to by-pass Windows security to alllow access in the way you describe. This is fundamental to UWP security model.

Are we talking about more than just NTFS permissions here?  I'm thinking that if we were to make the AppData a junction of a shared folder and reference the same one for multiple congregation user accounts, then the database would always be up to date as all congs would be using the same one!  That is unless the security you're describing is effected outside NTFS perms?

**Comment by AntonyCorbett** [comment]
@DanJermy I haven't tried to use junction points for this. I modify the default "Videos" and "Audio" folder for each user account and then in JWL Settings, ensure the correct folder is specified.

**Comment by DanJermy** [comment]
>  I modify the default "Videos" and "Audio" folder for each user account

@AntonyCorbett Sounds like that approach would achieve the same thing; JWL is effectively using the same location; however it still needs to be online to attempt to download a file which means the brothers in each cong still need to go through the process of 'getting' the media ready rather than it being in place because the previous cong last in had it ready for them.

**Comment by richfrantz** [comment]
I am just going to throw my 2 cents in here, but I would recommend keeping it as simple as possible. @DanJermy who in your Hall will be able to figure out all the work you did if you get hit by a bus tomorrow.
A unique account for each congregation. JWL using the default Music and Videos folder. Each congregation maintaining their own downloads.

**Comment by DanJermy** [comment]
@richfrantz Hi. Valid point, which is why I keep meticulous build documentation so anyone could replicate my efforts, and as much as possible is scripted to facilitate simple replication;  
It may surprise you but I am attempting to keep it simple.  Having each congregation help the next one out by having the downloads was a request from our brothers and seemed simple enough!

I felt that a single account was the simplest as suggested by several; it's only the recordings and media that I need to separate logically so brothers can manage those per congregation.

**Comment by DanJermy** [comment]
@AntonyCorbett Do you think that either of these are legitimate feature requests?...
- add a congregation-list selection box on launch like Soundbox did and marry that to a different working folder path
- allow a meeting time-window setting so that the correct congregation (and therefore working folder path) is correctly determined by the time/day.

Alternatively maybe I can use a batch file to retrieve the time and achieve an effective "IF" rule to pass the relevant congregation parameter as a switch to OnlyR and OnlyM?

Thanks for everyone's input.  I'm of the inclination that usually the default set up doesn't make the environment fool-proof so want to make it as simple as possible for our brothers to **use**; though I appreciate there's a little more effort in the **setup** in the first place.

**Comment by JohannesKreuzer** [comment]
@DanJermy I have written a small PowerShell script which shows as congregation selector and passes the name of the congregation to onlyR or onlyM. It does also apply a default confit of the app.

If some one is interested I’ll share the script.

**Comment by DanJermy** [comment]
@AntonyCorbett I'd love to try that please mate.  I'd just settled on a folder on the desktop with different shortcuts for each congregation, but this would be even better. Many thanks for your amazingly speedy help once again mate.

**Comment by AntonyCorbett** [comment]
> I'd love to try that please mate. 

@JohannesKreuzer from @DanJermy

**Comment by DanJermy** [comment]
@JohannesKreuzer apologies. In haste not noticed it wasn't Anthony replying. Many thanks for your help.

**Comment by LvDaniel** [comment]
In my opinion adding a congregation-list selection box on launch like in Soundbox it's te best way forward. And would be nice to have this for OnlyT and OnlyR as well.
If this is difficult to be implemented, then we have the script.
@JohannesKreuzer thanks for creating and sharing with us. Does a really great job.

**Comment by AntonyCorbett** [comment]
@LvDaniel what does the log indicate please?

**Comment by LvDaniel** [comment]
@AntonyCorbett Could not get metadata from file: C:\Users\Administrator\Documents\OnlyM\Media\M3-004 mwbv_M_201810_07_r720P.mp4 (in use)

**Comment by AntonyCorbett** [comment]
@LvDaniel Sorry I can't help with the script.

**Comment by LvDaniel** [comment]
@AntonyCorbett please ignore the script.
Do you know why when I try to start OnlyM from CMD using the command "C:\Program Files (x86)\OnlyM\OnlyM.exe" I get this behavior?
Is this related to OnlyM? For you it's working?
Thank you for your help and patience.

**Comment by JohannesKreuzer** [comment]
@LvDaniel I only tested the script with OnlyR and expected that it works w/o problems. I'll try to reproduce the issue.

For me it looks like a permission issue. Is the file also opened by another app? 
Are you working as user Adminstrator or is your user using the Media Directory of the user Administrator?

There seems to be a problem with getting the metadata of the file (System.IO.IOException).

@AntonyCorbett Is there a dependency on the working directory of OnlyM (the directory OnlyM is called from)?

@LvDaniel You could try to change the line
`& "C:\Program Files (x86)\OnlyM\OnlyM.exe" /id=$vers /nosettings`
to
`Set-Location "C:\Program Files (x86)\OnlyM"`
`& ".\OnlyM.exe" /id=$vers /nosettings`
so you don't need to have the scripts in the OnlyM directory.

**Comment by JohannesKreuzer** [comment]
@AntonyCorbett prefixing the ffmeg directory with `AppDomain.CurrentDomain.BaseDirectory` (as you use it to get the currently supported languages) would probably resolve the issue. 
For localized content I think you use the default .NET Culture mechanism, which is relying on the assembly loader for getting the resources.

**Comment by LvDaniel** [comment]
@JohannesKreuzer thanks, changing the line does the trick. Thanks for help and thanks also to @AntonyCorbett.

**Comment by AntonyCorbett** [comment]
@JohannesKreuzer Thanks for your help. I will experiment

**Comment by TizianoNoto** [comment]
I'm a bit late for the discussion but i'll post my experience in case it could be useful for someone.

Our KH hosts 3 congregations. The administrator chose to create _separate accounts_ (plus one for adminstration) and to instruct not to change hardware-related settings in the software we use. After upgrading to Windows 10 and installing JWL we opened its settings and _changed the media folders so that they point to the common (public) folders_.

The only problem we had with this setting happened yesterday. JWL wouldn't open nor re-download two videos. Turned out another congregation managed to re-download them with a different resolution (240p? Seriously?) so it probably was a file mismatch error. Always make sure **everybody download all videos with the same resolution** (use 720p for HD displays).

The only "complicated" thing we did was to create hard links to the already downloaded songs in two different folders so that both JW Library and TimePlayer actually see the same files.
